### PR TITLE
first comments

### PR DESCRIPTION
**Pull request by nlohmann**
A lot of small comments and edits. I also added some TODOs to discuss some things.

I have some more remarks beginning in the section where `push_back` is defined. I shall add them tomorrow.

### CODE REVIEW

**Comment by mariokonrad** [comment]
First: thank you for the spelling corrections, at least some mistakes are consistently ;-)

Discussion points:

1. Order of enumeration `json_type`: the old order was increasing complexity of the data. Similarity to python is also OK. Personally, I have no preference.

2. Equality of floating point numbers: maybe the equality could be formulated as the difference must be smaller or equal to `std::numeric_limits<T>::epsilon()`. The construction should catch `inf` and `NAN`, to prevent the necessity of checks in comparisons. They are forbidden by JSON anyway. Any thoughts?

3. `front()` and `back()`: indeed `begin()/end()` have constant complexity, so could `front()`. Depending on "special" implementations of containers, `back()` *could* be more complex than `end()`, e.g. trees. But since both, `front()` and `back()` rely on the underlying type, I think there cannot be other guarantees made by basic_json.

4. `clear()` to default construct with the same type: semantics, points could be made for both positions... Since the proposal is based on nlomann/json, let's go with it's behavior.

File: proposal.md [hunk 1]
```diff
@@ -2487,6 +2655,38 @@ template <> struct hash<json_v1::json>
 *Effect:* The template specialization to calculate the hash value for the specified
 JSON value.
 
+<a name="limitations"></a>
+## Limitations and Interoperability
```
Comment (mariokonrad) [inline proposal.md:2688]: This chapter was missing indeed.

The question remains: what are the limitations set by the proposal.

Regarding interoperability, I think it's the right choice to support UTF-8 but not UTF-16 and UTF-32 for the sake of simplicity.

File: proposal.md [hunk 2]
```diff
@@ -1369,25 +1481,27 @@ requirements:
...
-  ------------------------------------ | -------------
-  `json_type::null`                      | `0` (same as `size()`)
-  `json_type::boolean`                   | `1` (same as `size()`)
-  `json_type::string`                    | `1` (same as `size()`)
-  `json_type::number_integral_signed`    | `1` (same as `size()`)
-  `json_type::number_integral_unsigned`  | `1` (same as `size()`)
-  `json_type::number_floating_point`     | `1` (same as `size()`)
-  `json_type::object`                    | result of function `object_type::max_size()`
-  `json_type::array`                     | result of function `array_type::max_size()`
+| Value type                            | return value                                 |
```
Comment (mariokonrad) [inline proposal.md:1491]: This table renders strangely

File: proposal.md (hunk 1 above)
Comment (nlohmann) [reply proposal.md:2688]: Regarding Unicode: yes, UTF-8 is sufficient.

The rest: I don't know. I would leave the items as a reminder that if we make a limitation, we should acknowledge it.

**Comment by nlohmann** [comment]
Regarding https://github.com/nlohmann/std_json/pull/10#issuecomment-324148502:

1. Ok. (Removed the TODO)
2. I have a similar discussion (https://github.com/nlohmann/json/issues/703) right now. I'm not sure how much a wrapper around a floating-point number should mess with equality tests - especially, when it uses the `operator==` overload.
3. Ok. (Removed the TODO)
4. Ok. I shall make the changes.

File: proposal.md [hunk 3]
```diff
@@ -2003,12 +2003,16 @@ Example: `"cannot use swap() with boolean"`.
...
 
+##### Add element to array
+
 ```cpp
 void push_back(basic_json &&);
 void push_back(const basic_json &);
 void push_back(const typename object_type::value_type &);
 ```
 
+[TODO: Discuss. The third overload uses `object_type::value_type` which seems strange in the context of adding to an array.]
```
Comment (mariokonrad) [inline proposal.md:2013]: Yes, silly me. Most probably I missed to specify this particular overload in the context of object, silly me. Should be separated and specified accordingly.

**Comment by mariokonrad** [comment]
I suggest to merge and keep [nlohmann/json#703](https://github.com/nlohmann/json/issues/703) as an open issue.

**Pull request merged**